src/parser/index.ts

```diff
@@ -0,0 +1,178 @@
+import * as Expr from "./Expression";
+type Expression = Expr.Expression;
+import { Lexeme } from "../Lexeme";
+import { Token } from "../Token"
+import * as ParseError from "./ParseError";
+
+let current: number;
+let tokens: ReadonlyArray<Token>;
+
+export function parse(toParse: ReadonlyArray<Token>) {
```

Comment:
I definitely see this class just becoming a large function chain, but the fact that you have actually defined the functions in the order they are actually called makes it significantly easier to understand and follow. I'm sort of just thinking ahead, but will we ever want to separate out related parsing logic (e.g. expressions, for/while/for each loops, if/else, functions)? Or is the plan for now to sort of keep this class the home for all parsing logic? It is a bit hard to gauge on how hairy this class could become.

Reply:
That's a great point!  I've got no idea how rough this will become, but I suspect it'll just grow as a single file for quite a long time.  For comparison, [TypeScript's parser.ts](https://github.com/Microsoft/TypeScript/blob/0477f91dbad209809363323042f34926d3a6e175/src/compiler/parser.ts#L7498) weighs in at 7,498 lines `0_o`